src/document/generator.ml

```diff
@@ -1783,26 +1799,152 @@ module Make (Syntax : SYNTAX) = struct
+
+      and t_to_sidebar convert = function
+        | _, Leaf payload -> [ convert payload ]
+        | fallback, Dir d -> to_sidebar ~fallback convert d
+    end
+
+    let sidebar root_id (v : Odoc_model.Lang.Sidebar.t) =
+      let root_id = (root_id :> Paths.Identifier.t) in
+      let prepare (link_content, x) =
+        let x = (x :> Paths.Identifier.t) in
+        let payload =
+          (link_content, x, Odoc_model.Paths.Identifier.equal x root_id)
+        in
+        let path = Odoc_model.Paths.Identifier.fullname x in
+        (payload, path)
+      in
+      let title t =
+        block
+          (Inline [ inline (Inline.Styled (`Bold, [ inline (Inline.Text t) ])) ])
+      in
+      let page_hierarchy { Odoc_model.Lang.Sidebar.page_name; pages } =
+        if pages = [] then []
+        else
+          let pages = List.map prepare pages in
+          let hierarchy =
+            Hierarchy.make pages |> Hierarchy.remove_common_root
+          in
+          let convert (content, id, is_highlighted) =
+            let url = Url.from_identifier ~stop_before:false id in
+            match url with
+            | Ok href ->
+                let target = InternalLink.Resolved href in
+                let content = Comment.link_content content in
+                let link = { InternalLink.target; content; tooltip = None } in
+                let attr = if is_highlighted then [ "current_unit" ] else [] in
+                block (Inline [ inline ~attr @@ Inline.InternalLink link ])
+            | Error _ ->
+                let content = Comment.link_content content in
+                (* let attr = if is_highlighted then [ "current_unit" ] else [] in *)
+                block (Inline content)
+          in
+
+          let pages = Hierarchy.to_sidebar convert hierarchy in
+          let pages = [ block (Block.List (Block.Unordered, [ pages ])) ] in
+          [ title @@ page_name ^ "'s Pages" ] @ pages
+      in
+      let rec concat_map acc f = function
+        | hd :: tl -> concat_map (List.rev_append (f hd) acc) f tl
+        | [] -> List.rev acc
+      in
+      let page_hierarchies = concat_map [] page_hierarchy v.pages in
+      let units =
+        let item id =
+          let id = (id :> Paths.Identifier.t) in
+          let href, name, is_highlighted =
+            ( Url.from_identifier ~stop_before:false id,
+              Paths.Identifier.name id,
+              Odoc_model.Paths.Identifier.equal id root_id )
+          in
+          match href with
+          | Error _ -> None
+          | Ok href ->
+              let target = InternalLink.Resolved href in
+              let content = [ inline @@ Text name ] in
+              let link = { InternalLink.target; content; tooltip = None } in
+              let attr = if is_highlighted then [ "current_unit" ] else [] in
+              let elem =
+                [ block (Inline [ inline ~attr @@ Inline.InternalLink link ]) ]
+              in
+              Some elem
```

Comment:
Pages contain almost the same code.

Reply:
I factored what I could in the `sidebar_toc_entry` function in [4eb39d8](https://github.com/ocaml/odoc/pull/1145/commits/4eb39d8b11d1ea259ea8d7547fba2c871965aa6b) (and the previous commit).